sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
 
-sstable_set::incremental_selector::selection
-sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
+sstable_set::incremental_selector::incremental_selector(std::unique_ptr<incremental_selector_impl> impl, const schema& s, lw_shared_ptr<sstable_list> sstl)
+    : _impl(std::move(impl))
+    , _cmp(s)
+    , _sstl(std::move(sstl)) {
+}
+
+sstable_set::incremental_selector::selection sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
     if (!_current_range_view || !_current_range_view->contains(pos, _cmp)) {
-        std::tie(_current_range, _current_sstables, _current_next_position) = _impl->select(pos);
+        std::vector<shared_sstable> _current_versioned_sstables;
+        std::tie(_current_range, _current_versioned_sstables, _current_next_position) = _impl->select(pos);
+        _current_sstables = boost::copy_range<std::vector<shared_sstable>>(_current_versioned_sstables
+                             | boost::adaptors::filtered([this] (shared_sstable sst) { return _sstl->contains(sst); })
+                             | boost::adaptors::transformed([] (shared_sstable sst) { return sst; }));
         _current_range_view = _current_range->transform([] (const dht::ring_position& rp) { return dht::ring_position_view(rp); });
     }
     return {_current_sstables, _current_next_position};
 }
 
-sstable_set::incremental_selector
-sstable_set::make_incremental_selector() const {
-    return incremental_selector(_impl->make_incremental_selector(), *_schema);
+sstables::sstable_set::incremental_selector sstable_set::make_incremental_selector() const {
+    return incremental_selector(_all->version().make_incremental_selector(), *_all->version().get_schema(), _all);
+}
+
+sstable_set_version::~sstable_set_version() {
+    for (auto&& added_sst : _added) {
+        _versioned_data->remove(added_sst);
+    }
+    if (_prev) {
+        _prev->_next.erase(this);
+    }
+}
+
+sstable_set_version::sstable_set_version(std::unique_ptr<sstable_set_impl> impl, schema_ptr schema) 
+    : _versioned_data(make_lw_shared<versioned_sstable_set_data>(std::move(impl))) 
+    , _schema(std::move(schema)) {
+}
+
+sstable_set_version::sstable_set_version(sstable_set_version* ver) 
+    : _versioned_data(ver->_versioned_data)
+    , _schema(ver->_schema)
+    , _prev(ver) {
+        _prev->_next.insert(this);
+}
+
+size_t sstable_set_version::sstables_changed() const {
+    return _added.size() + _erased.size();
+}
+
+void sstable_set_version::merge_with_next() {
+    auto next_version = *_next.begin();
+    if (sstables_changed() > next_version->sstables_changed()) {
+        std::swap(_added, next_version->_added);
+        std::swap(_erased, next_version->_erased);
+    }
+    for (auto&& sst : _added) {
+        if (!next_version->_erased.contains(sst)) {
+            next_version->_added.insert(sst);
+            _versioned_data->insert(sst);
+        } else {
```

Comment:
This is called from a destructor, when a version is destroyed, yes? So it has to be noexcept and exception safe.

Maybe use container node operations (e.g. map::extract/map::merge)

Reply:
I've changed this and all other methods that are called from destructors to noexcept and exception safe using your advice (the node operations)